crates/burn-ndarray/src/ops/matmul.rs

```diff
@@ -11,101 +12,270 @@ pub(crate) fn matmul<E, const D: usize>(
 where
     E: FloatNdArrayElement,
 {
-    let shape_ori_lhs = lhs.shape();
-    let shape_ori_rhs = rhs.shape();
+    let shape_lhs = lhs.shape();
+    let shape_rhs = rhs.shape();
+    let m = shape_lhs.dims[D - 2]; // # of left rows
+    let k = shape_rhs.dims[D - 2]; // # of left cols and right rows
+    let n = shape_rhs.dims[D - 1]; // # of right cols
 
-    let lhs = reshape(lhs);
-    let rhs = reshape(rhs);
+    let (out_shape, strides_lhs, strides_rhs, strides_out) = output_shape(&shape_lhs, &shape_rhs);
+    let l_mat_size = m * k; // size of matrix component of left array
+    let r_mat_size = k * n; // size of matrix component of right array
+    let out_mat_size = m * n; // size of matrix component of output array
 
-    let [batch_size_lhs, m, _] = lhs.shape().dims;
-    let [batch_size_rhs, _, n] = rhs.shape().dims;
+    let num_l_batches = shape_lhs.num_elements() / l_mat_size;
+    let num_r_batches = shape_rhs.num_elements() / r_mat_size;
+    let num_out_batches = out_shape.num_elements() / out_mat_size;
 
-    let mut shape_out = match batch_size_lhs > batch_size_rhs {
-        true => shape_ori_lhs,
-        false => shape_ori_rhs,
-    };
-    shape_out.dims[D - 2] = m;
-    shape_out.dims[D - 1] = n;
+    let alpha: E = 1.0.elem();
+    let beta: E = 0.0.elem();
 
-    let out = general_matmul(lhs, rhs);
-
-    NdArray::<E>::float_reshape(out, shape_out)
-}
-
-fn general_matmul<E: FloatNdArrayElement>(
-    lhs: NdArrayTensor<E, 3>,
-    rhs: NdArrayTensor<E, 3>,
-) -> NdArrayTensor<E, 3> {
-    run_par!(|| {
-        let [batch_size_lhs, m, _] = lhs.shape().dims;
-        let [batch_size_rhs, k, n] = rhs.shape().dims;
-        let batch_size = usize::max(batch_size_rhs, batch_size_lhs);
-
-        if batch_size_lhs > batch_size && batch_size_lhs != 1 {
-            panic!("Broadcast on multiple dimensions is not yet supported");
-        }
-
-        if batch_size_rhs > batch_size && batch_size_rhs != 1 {
-            panic!("Broadcast on multiple dimensions is not yet supported");
-        }
-
-        let alpha: E = 1.0.elem();
-        let beta: E = 0.0.elem();
-
-        let mut out_array = ndarray::Array3::<E>::zeros((batch_size, m, n));
+    let out: NdArrayTensor<E, D> = run_par!(|| {
+        let mut out_array = ndarray::Array3::<E>::zeros((num_out_batches, m, n));
         let unsafe_shared_out_array = UnsafeSharedRef::new(&mut out_array);
 
-        let lhs_array = lhs.array.into_shape((batch_size_lhs, m, k)).unwrap();
-        let rhs_array = rhs.array.into_shape((batch_size_rhs, k, n)).unwrap();
+        let lhs_array = NdArray::<E>::float_reshape(lhs, Shape::new([num_l_batches, m, k])).array;
+        let rhs_array = NdArray::<E>::float_reshape(rhs, Shape::new([num_r_batches, k, n])).array;
```

Comment:
I originally tried this with just `into_shape`, but that fails in some cases with an error about an incompatible layout. The original code was also calling `float_reshape` (below), so I've re-used that here.